### PR TITLE
kms: Track used devices and free resources

### PR DESCRIPTION
**Pull request by Drakulix**
The goal of this PR is to track which device we need to have active EGL resources for and to get rid of those to allow dedicated gpus to enter deep sleep.

The metric right now is:
- ~~We need to have the primary device open to import dmabufs from the main dmabuf feedback~~ Theoretically we could now even stop the primary gpu, if it has no outputs, but it remains a fallback for rendering in some scenarios (screencopy), so we need to carefully test this to really shut it down.
- We have to have gpu's with attached outputs open to be able to swap to them as compositing devices (they will use power anyway and keeping the GL context alive won't prevent them from going to a low-power state)
- ~~We need to have GPUs open, when client use explicit sockets (as that GPU will be advertised as a main device for dmabuf feedback) or when using the `COSMIC_RENDER_AUTO_ASSIGN` variable for dmabuf imports. (In theory the latter is redundant as these devices do have attached outputs, but those might be disconnected during the lifetime of the client, making it necessary to track the clients.)~~ Now tracks buffers that are actually imported on a given GPU.
  - ~~TODO: Track used GPU nodes and reintroduce `render_node_for_output`. Also remove GPUs on buffer free again~~ done.

### CODE REVIEW

**Comment by Drakulix** [comment]
Now based on https://github.com/pop-os/cosmic-comp/pull/292.

Tries all kms-nodes for dmabuf import again to facilitate screencopy from different gpu for https://github.com/pop-os/xdg-desktop-portal-cosmic/issues/19

File: src/backend/kms/mod.rs [hunk 1]
```diff
@@ -1010,49 +1061,58 @@ impl Device {
...
-        return xwayland_client.user_data().get::<DrmNode>().cloned();
-    }
-    None
+}
+
+fn source_node_for_surface<'a>(
+    w: &WlSurface,
+    mut devices: impl Iterator<Item = &'a Device>,
+) -> Option<DrmNode> {
+    with_renderer_surface_state(w, |state| {
```
Comment (ids1024) [inline src/backend/kms/mod.rs:1074]: I'm seeing a panic here in the `unwrap` within `with_renderer_surface_state` when I start an X client.

Comment (Drakulix) [reply src/backend/kms/mod.rs:1074]: https://github.com/Smithay/smithay/pull/1317

**Comment by ids1024** [comment]
Odd, I'm seeing some kind of artifacting (like parts of the screen black for a frame?) with an Intel-rendered Firefox window on the external monitor connected to a 1650 mobile.

It seems I can reproduce this consistently while scrolling, with Firefox as the only thing open on that output, in https://github.com/pop-os/cosmic-comp/pull/278/commits/badb8ed12f128e5843b67d5455d6e62ebffa5889 but not before that.

**Comment by Drakulix** [comment]
Draft while this depends on unreleased smithay changes

**Comment by ids1024** [comment]
I guess stride issues are still an issue with Nvidia....

DP-4

This is an intel surface being dragged on an Nvidia output.

There also seems to be an issue for frame callbacks of drag surfaces. If I drag the window to a different monitor, it drops to 1fps until I release the drag.

**Comment by ids1024** [comment]
With Intel surfaces on my Nvidia monitor, that don't fill the screen, I'm seeing some kind of flickering dot grid in part of the background? If you look at the top left corners of the windows, and look further up from there.

PXL_20240215_173125841 1

It seems to be damage-related, since moving a window over that part of the screen clears it.

**Comment by ids1024** [comment]
With es2gears rendered by Intel, filling most of the 1440p screen connected to the Nvidia monitor, I seem to see a higher frame rate when the window is being dragged... but only when built with the `debug` feature? When the window is dragged it composites with the Nvidia GPU instead of the Intel GPU, but I don't know what `debug` would change...

With both the `debug` feature enabled, and the window being dragged, it gives an fps a bit under 30. Naturally it should be 60, but this is similar to Gnome on this 1650 card.

**Comment by ids1024** [comment]
Running a game application rendered with the Nvidia GPU (with `SDL_VIDEODRIVER=wayland WAYLAND_DISPLAY=wayland-1-renderD129`) on the Intel monitor fails to show correctly:

```
Feb 15 12:56:13 pop-os cosmic-comp-debug[3467754]: [EGL] 0x3009 (BAD_MATCH) eglCreateImageKHR: createImageFromDmaBufs failed
Feb 15 12:56:13 pop-os cosmic-comp-debug[3467754]: Error loading buffer: Error on the rendering device: Unknown pixel format
Feb 15 12:56:13 pop-os cosmic-comp-debug[3467754]: Failed to import surface: Error on the rendering device: Unknown pixel format
```

It's creating buffers like:

```
[3801939.941]  -> zwp_linux_dmabuf_v1@31.create_params(new id zwp_linux_buffer_params_v1@43)
[3801939.961]  -> zwp_linux_buffer_params_v1@43.add(fd 31, 0, 0, 10240, 50331648, 6316052)
[3801939.968]  -> zwp_linux_buffer_params_v1@43.create_immed(new id wl_buffer@44, 2560, 1440, 875713112, 0)
[3801939.974]  -> zwp_linux_buffer_params_v1@43.destroy()
```

If that window is on the Nvidia-connected output though, it runs fullscreen at the monitors refresh rate of 60fps just fine.

**Comment by flukejones** [comment]
```
smithay = {git = "https://github.com/smithay//smithay", branch = "multigpu-shadow-copy"}
[patch.crates-io]
gbm = {git = "https://github.com/smithay/gbm.rs", branch = "multigpu-rework"}
```

these branches seem to be missing?

**Comment by ids1024** [comment]
The `multigpu-shadow-copy` branch exists, but the commit in the lockfile doesn't. The gbm change is merged and released.

It should build if you run `cargo update -p smithay`. And change `rust-toolchain.toml` to `1.72`.

**Comment by flukejones** [comment]
@ids1024 that worked, thanks. Looks like good progress. I continuously test on Nvidia output to a 5k ultrawide.

**Comment by Drakulix** [comment]
> Running a game rendered with the Nvidia GPU (with `SDL_VIDEODRIVER=wayland WAYLAND_DISPLAY=wayland-1-renderD129`) on the Intel monitor fails to show correctly:
> 
> ```
> Feb 15 12:56:13 pop-os cosmic-comp-debug[3467754]: [EGL] 0x3009 (BAD_MATCH) eglCreateImageKHR: createImageFromDmaBufs failed
> Feb 15 12:56:13 pop-os cosmic-comp-debug[3467754]: Error loading buffer: Error on the rendering device: Unknown pixel format
> Feb 15 12:56:13 pop-os cosmic-comp-debug[3467754]: Failed to import surface: Error on the rendering device: Unknown pixel format
> ```
> 
> It's creating buffers like:
> 
> ```
> [3801939.941]  -> zwp_linux_dmabuf_v1@31.create_params(new id zwp_linux_buffer_params_v1@43)
> [3801939.961]  -> zwp_linux_buffer_params_v1@43.add(fd 31, 0, 0, 10240, 50331648, 6316052)
> [3801939.968]  -> zwp_linux_buffer_params_v1@43.create_immed(new id wl_buffer@44, 2560, 1440, 875713112, 0)
> [3801939.974]  -> zwp_linux_buffer_params_v1@43.destroy()
> ```

This was probably related the opaque format of that buffer. The copy code contained a missing conversion for that case, that should be solved on that smithay branch now.

Additionally copying from external textures should now also be fixed.
(This can happen in the very stupid case of a linear buffer from the intel-gpu imported by the nvidia-gpu being displayed on the intel-gpu... wgpu does this when set to low-power and the nvidia card is set as the primary device.)

**Comment by Drakulix** [comment]
> I know this has already been merged... but it umm breaks this one native Linux game. :D I know... I could buy it for you if you wanna test it yourself.

Which one? And how does it break?

**Comment by Vixea** [comment]
Game called Path of Titans. The game doesn't receive input and only updates around the cursor or resizes.

**Pull request merged**